docs/design/features/StringDeduplication.md

```diff
@@ -0,0 +1,226 @@
+
+- We know when new LOH strings are allocated (all LOH allocations are done via C++ helpers in gchelpers.cpp), we can take the accumulative size of new LOH strings till it reaches a certain percentage of LOH before starting the next dedup cycle.
+
+- We know how much is promoted into gen2 and we can also wait till a certain percentage of gen2 consists of newly promoted memory before starting the next dedup cycle.
+
+#### **Compatibility**
+
+The following scenarios become problematic or more problematic when deduping is on so we need ways to help users find them.
+
+- Mutating the string content
+
+Strings are supposed to be immutable. However in unsafe code you can change the string content after it’s created. Changing string content already asking for trouble without deduping – you could be changing the interned copy which means you are modifying someone else’s string which could cause completely unpredictable results for them. 
+
+The most common way is to use the fixed keyword:
+
+```c#
+fixed (char* p = str)
+{
+  p[0] = 'C';
+}
+```
+
+There are other ways such as 
+
+`((char*)(gcHandlePointingToString.AddrOfPinnedObject())) = 'c';`
+
+Or
+
+`Unsafe.AsRef (in str.GetPinnableReference()) = 'c';`
+
+- Locking on a string
+
+Locking on a string object is already discouraged due to a string can be interned. Having string dedup on can make this problematic more often if the string you called lock on is now deduped to a different string object. 
+
+- Reference equality
+
+This example illustrates an example of `ReferenceEquals` that will not work if `_field` was deduped to a different string reference by the time the assignment to a2 happens.
+
+```c#
+string _field;
+
+void Test()
+{
+    _field = new string('a', 1);
+
+    string a1 = _field;
+    Thread.Sleep(1000);
+    string a2 = _field;
+
+    // This assert can fail with string deduping turned on
+    Debug.Assert(ReferenceEquals(a1,a2));
+}
+```
+
+**Runtime assistance to help with these scenarios**
+
+- Static analysis
+
+For .NET 5.0, there’s a static analysis [plan](https://github.com/dotnet/runtime/issues/30740). Verifying if string content is mutated can be part of that analysis suite. It would also be beneficial to have VS do this verification and give you a warning when you are mutating string content.
+
+To start with we will provide analysis for the following –
+
+1. Using the fixed keyword to modify string content.
+
+I’m seeing that there are almost 600 places in libraries that do `fixed (char*` but hopefully most of them do not actually modify the string content. We should definitely be encouraging folks to switch to using `string.Create` like what [PR#31700](https://github.com/dotnet/runtime/pull/31700) did.
```

Comment:
We use this pattern internally for implement string constructors. We will need an internal API that will be called at the end of the string constructor to mark the string as "baked" and candidate for deduping.

Reply:
oh right, I should add this...sorry I forgot about it.

Reply:
Actually, we won't need this method if we are not deduping pointers on stack as @benaadams pointed out. Now I remember we talked about it already, but I forgot. Either way, it would be useful to mention it in the doc.

Reply:
ahh, because it can not be assigned to any field before it gets full constructed. makes sense.